internal/pkg/chain/sampler.go

```diff
@@ -0,0 +1,80 @@
+package chain
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/minio/blake2b-simd"
+
+	"github.com/filecoin-project/go-filecoin/internal/pkg/block"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/crypto"
+)
+
+// A sampler draws randomness seeds from the chain.
+type Sampler struct {
+	reader TipSetProvider
+}
+
+func NewSampler(reader TipSetProvider) *Sampler {
+	return &Sampler{reader}
+}
+
+// Draws a randomness seed from the chain identified by `head` and the highest tipset with height <= `epoch`.
+// If `head` is empty (as when processing the genesis block), the seed is empty.
+func (s *Sampler) Sample(ctx context.Context, head block.TipSetKey, epoch abi.ChainEpoch) (crypto.RandomSeed, error) {
+	var ticket block.Ticket
+	if !head.Empty() {
+		start, err := s.reader.GetTipSet(head)
+		if err != nil {
+			return nil, err
+		}
+		// Note: it is not an error to have epoch > start.Height(); in the case of a run of null blocks the
+		// sought-after height may be after the base (last non-empty) tipset.
+		// It's also not an error for the requested epoch to be negative.
+
+		tip, err := s.findTipsetAtEpoch(ctx, start, epoch)
+		if err != nil {
+			return nil, err
+		}
+		ticket, err = tip.MinTicket()
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Sampling for the genesis block.
+		ticket.VRFProof = []byte{}
+	}
```

Comment:
I think this might be wrong. Specifically, the genesis block will have an explicit genesis ticket. This does mean that `findTipsetAtEpoch` should be special cased so as to return the genesis tipset for any epoch `e <= 0`

Reply:
Yeah, my understanding from Lotus' plan which seems good to me is that this will be the same indeed (ie the initial ticket value, which will be used for the first LE). That value will be specified closer to launch. You can coordinate with Lotus on what to use for Painnet, but best have it as a placeholder either way.